Podcast Player: Add block settings

retrofox opened the pull request:
This PR adds the following setting sections to the block:

* Block Stypes: it adds `small` (default) and `large` styles
* Feed URL: refact the code creating a `<FeedUrlControl />` to be used in both block itself and sidebar.
* Add `Show Cover Art` and `Show Episode Description` toggle elements.

The colors section will be tackled in another PR.

**Depends on https://github.com/Automattic/jetpack/pull/14948**
Fixes #14940 (partially)

### Design

<img width="1115" alt="Screen Shot 2020-03-13 at 12 46 33 PM" src="https://user-images.githubusercontent.com/77539/76637014-c798d280-6528-11ea-9a4d-bc6072207fc0.png">

#### Testing instructions:
* Create/edit podcast player.
* Check that you can edit the block styles, Feed URL, and the _cover art_ and _episode description_ status
* Confirm that these changes are persistent. Hard refresh and they still should be there.

<img width="957" alt="Screen Shot 2020-03-17 at 10 18 29 AM" src="https://user-images.githubusercontent.com/77539/76859796-be657980-6838-11ea-9c65-decdf90b841e.png">

### Proposed changelog entry for your changes:

*

retrofox committed:
[not verified] podcast-player: move supportLink outside

retrofox committed:
[not verified] podcast-player: add block styles

retrofox committed:
[not verified] podcast-player: refactoring -> add FeedURLControl
It isolates a new component called FeedURLControl, since it will be reused in the Inspector Control soon.

retrofox committed:
[not verified] podcast-player: change the Feed from the sidebar

retrofox committed:
[not verified] podcast-player: add label to Feed url control

retrofox committed:
[not verified] podcast-player: tweak FeedURLControl component

retrofox committed:
[not verified] podcast-player: useCallback for checkPodcastLink

getdave commented:
I wonder whether you should have to click edit to change the feed url in the Inspector Controls?

retrofox commented:
> I wonder whether you should have to click edit to change the feed url in the Inspector Controls?

What `Edit` do you mean? 😅 AFAIK, all changes done by controls are immediately propagated.

retrofox committed:
[not verified] podcast-player: add toggle setting options

jeherve reviewed (changes requested):
This one will need a rebase as well.

retrofox commented:
> This one will need a rebase as well.

✅

obenland commented on extensions/blocks/podcast-player/index.js:
```diff
@@ -24,6 +24,12 @@ export const settings = {
+	styles: [
+		{ name: 'small', label: __( 'Small' ), isDefault: true },
+		{ name: 'large', label: __( 'Large' ) },
+	],
```
Are these functional?

retrofox replied:
Yes, but the styles aren't implemented yet. You can select the styles from the sidebar as well.

<img src="https://user-images.githubusercontent.com/77539/76799338-10ae8800-67b0-11ea-83a9-274c9546478d.png" width="500px" />

obenland replied:
Could we either hold off on that until we have styles or add styles to make it functional?

getdave replied:
...and it the styles add a load of extra work not directly related to "settings" then we could consider doing this in a follow up PR.

retrofox replied:
> Could we either hold off on that until we have styles or add styles to make it functional?

I think it's part of the development process. This is not the only change that will keep in a middle state among all PRs/patches that we are doing day by day. The scope of this PR is adding the block settings in the sidebar.

> ...and it the styles add a load of extra work not directly related to "settings" then we could consider doing this in a follow up PR.

Why do we need to discard something that we are going to use in the near future? I prefer to discard it in the future (if something unexpected happens) than now.
I've based the progress of this PR according to the [given design](https://github.com/Automattic/Dotcom-roadmap/issues/100). The lack of designs is something that we've lamented in previous iterations.

I prefer to keep these changes here and keep iterating in other patches.

obenland committed:
[not verified] [not verified] Make codeclimate happy

obenland committed:
[not verified] [not verified] Make # of items work.

obenland commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -35,14 +41,53 @@ const handleSSRError = () => {
 };
 
+// Support page link.
+const supportLink =
```
`supportUrl`?

obenland commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -35,14 +41,53 @@ const handleSSRError = () => {
+		onKeyPress={ ev => ( ev.key === 'Enter' ? onEnterKeyPress( ev ) : noop ) }
+		onBlur={ onBlur }
+	/>
+);
```
I'm not sure I'm too big of a fan of this abstraction. It doesn't seem to save us much work and obfuscates the underlying component. What do you think?

getdave replied:
I'm interested to understand _why_ this abstraction was added. At first glance I can't see any major benefit but @retrofox may well have a good reason that we're not aware of.

getdave replied:
Looking at the rest of the changes I support this component abstraction. You will notice that `FeedURLControl` is used twice in the component

1. Block Controls
2. Inspector Controls

It makes sense to DRY this up especially given the number of props being applied. It would be easier to change one and forget the other. Having a single component helps avoid this.

retrofox replied:
> I'm not sure I'm too big of a fan of this abstraction.

It's part of the implementation. The idea is to reuse the same component in two places.

>  It doesn't seem to save us much work and obfuscates the underlying component. What do you think?

Disagree. ;-)
The component takes over updating the Podcast Link. It expects three arguments and 💥 you get the input filed ready to use. Glad to update the doc if we consider appropriate.

> Looking at the rest of the changes I support this component abstraction. You will notice that FeedURLControl is used twice in the component

It's already documented in the function :-D :

```es6
/**
 * Control component used either inside of the Podcast block
 * as well as into the InspectorControls section (sidebar)
 *
 * @param {string}      url             Podcast Feed URL
 * @param {string|Null} label           Input field label. Optional.
 * @param {Function}    onUrlChange     onChange text handler.
 * @param {Function}    onEnterKeyPress on Enter key input event handler. Optional.
 * @param {Function}    onBlur          onBlur input handler. Optional.
 * @return {*}                          React component
 */
```

Maybe it' unclear or incomplete. Happy to improve it :-) 

> It makes sense to DRY this up especially given the number of props being applied. 

Agree.

Thanks, @obenland and @getdave for this feedback. My suggestion is keeping this component as it, maybe improving the jsdoc. But glad to revert it if we consider good, and _kind of_ duplicate the code un both places where the component is consumed.

obenland commented:
@lessbloat Testing this PR, I noticed the design calling for an input field in the sidebar to edit the feed url. The RSS block that we've based the Podcast Player block on has an edit link in the block toolbar, which feels a little more natural to me:

<img width="167" alt="image" src="https://user-images.githubusercontent.com/1398304/76812261-89621400-67b1-11ea-8a0e-c88d283dee57.png">


How would you feel about letting the "content" of the block to be edited in the block itself and keeping the sidebar for presentational settings only?

jeherve commented on extensions/blocks/podcast-player/podcast-player.php:
```diff
@@ -60,7 +68,7 @@ function render_block( $attributes ) {
 	$attributes['url'] = esc_url_raw( $attributes['url'] );
 
-	$track_list = get_track_list( $attributes['url'], 10 );
+	$track_list = get_track_list( $attributes['url'], $attributes['itemsToShow'] );
```
I would ensure we always pass a proper integer here.

```suggestion
	$track_list = get_track_list( $attributes['url'], absint( $attributes['itemsToShow'] ) );
```

getdave replied:
Let's just generally make a rule not to trust attributes at all.

getdave commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -56,7 +101,7 @@ const PodcastPlayerEdit = ( {
 	 * @param {object} event Form on submit event object.
 	 */
-	const checkPodcastLink = event => {
+	const checkPodcastLink = useCallback( event => {
```
Minor: we [probably don't need `useCallback`](https://kentcdodds.com/blog/usememo-and-usecallback) here. As with `useMemo` the primary use cases for this are:

There are specific reasons both of these hooks are built-into React:

1. Referential equality
2. Computationally expensive calculations

I don't think we have either of these here so if anything it's probably adding extra overhead to call the `useCallback` hook.

retrofox replied:
Yeah, good point. I've seen that we use callback also when the bound function is tied to an event, like as in this case, `onSubmit`. Let's keep this comment as `unresolved` for now.

getdave commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -92,13 +132,7 @@ const PodcastPlayerEdit = ( {
-						onChange={ setEditedUrl }
-						className={ 'components-placeholder__input' }
-					/>
+					<FeedURLControl url={ editedUrl || '' } onUrlChange={ setEditedUrl } />
```
```suggestion
					<FeedURLControl url={ editedUrl  } onUrlChange={ setEditedUrl } />
```

Could we set defaults within the component?

retrofox replied:
Yeah, in fact, [it already does it](https://github.com/Automattic/jetpack/pull/14973/files#diff-ac95db14f00bcb6ae1960269ae9121cfR72).

getdave commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -127,6 +161,14 @@ const PodcastPlayerEdit = ( {
 				<PanelBody title={ __( 'Podcast settings', 'jetpack' ) }>
+					<FeedURLControl
+						label={ __( 'RSS Feed URL', 'jetpack' ) }
+						url={ editedUrl || '' }
```
```suggestion
						url={ editedUrl  }
```

As above, should we set defaults within the component?

getdave commented on extensions/blocks/podcast-player/podcast-player.php:
```diff
@@ -113,11 +124,11 @@ function render_player( $track_list, $attributes ) {
- * @param int    $quantity the number of tracks to return.
- * @return array|WP_Error the feed's tracks or a error object.
+ * @param string $feed     The RSS feed to load and list tracks for.
+ * @param int    $quantity Optional. The number of tracks to return. Default 5.
```
```suggestion
 * @param int    $quantity Optional. The number of tracks to return.
```

Minor: not sure we need the default in the description 🤷‍♂ Another thing to maintain.

getdave reviewed (dismissed):
Thanks for your work on this. I tested it.


I wasn't able to visually verify any of the changes. Therefore I used "Code editor" mode to do this (suggestion: we should update the testing instructions to make the process and the expectations clearer).



### Editor 
* Switching any setting toggle to `false` caused the attribute to be set to `false`. <img width="720" alt="Screenshot 2020-03-17 at 10 21 00" src="https://user-images.githubusercontent.com/444434/76846920-2ddb6900-6839-11ea-8807-f2d69defb831.png">
 * Switching any setting toggle to `true` caused the attribute to be missing from the attributes in the code view. <img width="769" alt="Screenshot 2020-03-17 at 10 30 39" src="https://user-images.githubusercontent.com/444434/76847701-62035980-683a-11ea-9613-944b43a66c5a.png">
* Toggle the styles caused an appropriate class name to be applied.

## Frontend

* Unable to verify description or cover art settings were applied. Not sure how to do this.
* Verified that classnames are applied correctly based on the the "style" setting.

# Bug

I notice an issue with the RSS feed URL in the Inspector controls. If using keyboard, blurring out of the field via `TAB` doesn't cause the setting to update. Therefore you can end up with a strange state where the field reads `s` and the Block shows the previous RSS feed URL:

![Screen Capture on 2020-03-17 at 10-26-38](https://user-images.githubusercontent.com/444434/76847483-09cc5780-683a-11ea-9da9-93fa845595ae.gif)

Also I wonder whether, given how important the feed URL setting is, we should require the following process to edit it:

1. Click "Edit button"
2. Field becomes editable.
3. Change URL.
4. Hit submit.
5. Validation (will not update without "passing" validation).
6. URL updated.
7. Block UI updates

This will protect users from themselves as it's too easy to accidentally modify the URL as it stands.

retrofox committed:
[not verified] podcast-player: remore fallback in prop

It removes the fallback value from the component prop since it's already handled by the component in itself.

Co-Authored-By: Dave Smith <getdavemail@gmail.com>

retrofox committed:
[not verified] podcast-player: minor doc improvement

Remove default value from jsdoc

Co-Authored-By: Dave Smith <getdavemail@gmail.com>

retrofox committed:
[not verified] podcast-player: remore fallback in prop

It removes the fallback value from the component prop since it's already handled by the component in itself.

Co-Authored-By: Dave Smith <getdavemail@gmail.com>

retrofox committed:
[not verified] poscast-player: cast itemsToShow

Ensure we always pass a proper integer for this attribute

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

retrofox commented:
> I wasn't able to visually verify any of the changes.

It's doable just setting the block properties, make a hard refresh, and confirm that the changes are persistent.

> Therefore I used "Code editor" mode to do this

This is a very nice way to see in detail what happens with the attributes. 👍 

> (suggestion: we should update the testing instructions to make the process and the expectations clearer).

I consider that setting the attribute, save the post, hard refresh and confirm that attributes values are persistent is enough for the testing purposes. his is how all blocks behave.

retrofox commented:
> I notice an issue with the RSS feed URL in the Inspector controls. If using keyboard, blurring out of the field via TAB doesn't cause the setting to update. 

Weird and nice catch. It takes my attention since we are listening to the `blur` event on the input element. Looking.

retrofox committed:
[not verified] podcast-player: rename to supportUrl

retrofox commented:
> > I notice an issue with the RSS feed URL in the Inspector controls. If using keyboard, blurring out of the field via TAB doesn't cause the setting to update.
> 
> Weird and nice catch. It takes my attention since we are listening to the `blur` event on the input element. Looking.

The input field propagates the change properly. The issue, in this case, is that `s` is not a valid URL, so it won't reflect it in the block. We need to show the error in the UI.

<img src="https://user-images.githubusercontent.com/77539/76858417-29618100-6836-11ea-84b6-ad4d1ab68afd.png" width="400px" />

retrofox commented:
> @lessbloat Testing this PR, I noticed the design calling for an input field in the sidebar to edit the feed url. The RSS block that we've based the Podcast Player block on has an edit link in the block toolbar, which feels a little more natural to me:
> 
> <img alt="image" width="167" src="https://user-images.githubusercontent.com/1398304/76812261-89621400-67b1-11ea-8a0e-c88d283dee57.png">
> 
> How would you feel about letting the "content" of the block to be edited in the block itself and keeping the sidebar for presentational settings only?

It makes sense considering that similar blocks don't add an input field to edit the URL.

**RSS**
<img src="https://user-images.githubusercontent.com/77539/76859359-ed2f2000-6837-11ea-9b1c-73c4f1c6f807.png" width="400px" />

**Youtube**
<img src="https://user-images.githubusercontent.com/77539/76859367-f61ff180-6837-11ea-98cc-d34523e269ce.png" width="400px" />

marekhrabe commented:
I would vote for removing from the sidebar. You can see from the code how much it increases the general complexity of this block - suddenly we have to validate several different ways (submit, blur, enter, …) and we might have inconsistent error reporting, as the URL can be edited in multiple places. Having a single one would simplify things and is already working well

retrofox commented:
> I would vote for removing from the sidebar. You can see from the code how much it increases the general complexity of this block - suddenly we have to validate several different ways (submit, blur, enter, …) and we might have inconsistent error reporting, as the URL can be edited in multiple places. Having a single one would simplify things and is already working well

I agree with removing it but disagree about the reasons :-D
Dealing with these issues is perfectly doable. In fact, the current implementation does it with the exception of showing the error message when it  isn't a valid URL. Styling with an error color or something like that should be enough, imo.

<img src="https://user-images.githubusercontent.com/77539/76861929-1ce02700-683c-11ea-81b1-57e59582aac0.png" width="400px" />

davemart-in commented:
> How would you feel about letting the "content" of the block to be edited in the block itself and keeping the sidebar for presentational settings only?

Sorry I'm late to the convo. This is fine, but Matias shared with me that the pencil icon turned out to be confusing for people. One alternative example he gave is where we added "Replace" in the image block:
![image](https://user-images.githubusercontent.com/5634774/76874678-e9a29580-6845-11ea-87bb-35796e05edd4.png)

My vote would be to follow suit and replace the pencil icon with the word "Update".

Open to your guys thoughts.

retrofox committed:
[not verified] podcast-player: remove RSS link from sidebar

retrofox committed:
podcast-player: restore items to show to 10

jeherve reviewed (approved):
This looks good to me. It should be good to merge!

retrofox commented:
r204411-wpcom
